### ISSUE TITLE
Implement flake8-walrus

### ISSUE THREAD

**Issue by Cielquan** [charliermarsh/ruff#3464]
# [flake8-walrus](https://pypi.org/project/flake8-walrus/)

Forbids usage of the `:=` walrus operator.

### Error Codes
* [ ] `ASN001` do not use assignment expressions

**Comment by kyoto7250** [comment]
Personally, I would like to work on implementing this plugin, but is this an appropriate rule to incorporate into `ruff`?

I know there were objections to the introduction of this operator, but I feel this rule is a matter of user preference.

**Comment by MichaReiser** [comment]
What's the rational of forbidding the operator (other than `lol` :D). Is it useful to a broad audience of the Python community?

**Comment by Cielquan** [comment]
The objection is readability. It may not be the only one, but it is mine.
I totally see that it is somewhat of a nice QoL feature, but I and others too do not want to exchange readability for 1 line of code saved.

And I do not want others to be able to use it in code I maintain, so I forbid its usage via the flake8 plugin by running flake8 in CI.

It is totally subjective preference but so is any style rule somehow, is it not?
But in comparison to PEP8 it is rather niche I guess. So I can see why one would not want to implement it. But nontheless I and probably others too would like to have it being implemented. 😉

**Comment by charliermarsh** [comment]
Yeah, I think this is a hard one for us to merge until we have a clearer policy on opinionated rules / plugins (similar to #3463).

**Comment by MichaReiser** [comment]
> It is totally subjective preference but so is any style rule somehow, is it not?
But in comparison to PEP8 it is rather niche I guess. So I can see why one would not want to implement it. But nontheless I and probably others too would like to have it being implemented.

We're considering reorganizing the ruff rules and one category that we intend to introduce is `restriction`. Rules in the `restriction` category are opinionated, not recommended for most users, and restrict the allowed language feature. I could see this rule fit into that category if it satisfies the criteria that it is useful to some users of the community (we don't want to have "I don't like" rules, where *I* is a single person).

**Comment by Cielquan** [comment]
Well the walrus operator is very dividing. There are people who like it and people who don't. It's a matter of taste after all I think.

I see value in have an optional restriction rule to forbid it's usage, but I also see value in an optional rule like auto-walrus.

I personally used it once in a list comprehension where it made the code simpler in my opinion. But in a general use like in your example I personally don't like it, because I think the code without is easier to read. But that is just my taste and I respect your taste if you like to use it in your code base.

**Comment by Avasam** [comment]
> https://github.com/astral-sh/ruff/issues/13741 is related where the author asked for restricting most, but not all walrus operators.

I do feel like that was a different ask. Removing redundant walrus leads to the same come but with 1 less character and autofixable.

Not the same as forbidding/forcing walrus usage where the code would need to be changed by adding or removing a separate assignement.

**Comment by k0pernikus** [comment]
> What's the rational of forbidding the operator (other than `lol` :D). Is it useful to a broad audience of the Python community?

I just banned it in one of my projects and am currently in the middle of writing a custom ast-grep linter to catch it mechanically right now, which is why I looked here first to see if it was already implemented so I don't have to.

Or more broadly: I banned the walrus operator (`:=`) *and* inline mutation generally because I want a hard line between statements and expressions: statements act (bind names, change state); expressions only compute a value. Data flow should reads top-to-bottom. A name is defined on its own line, above its use, never randomly mid-expression. `n = len(data)` then `if n > 10:`, not `if (n := len(data)) > 10:`.

It adds too much mental load, and esp. if the if-branches are side-effect prone, it gets even more trickier to parse.

Esp. this:

```python
class Probe:
    def __init__(self) -> None:
        self.f_runs = 0
        self.g_runs = 0

    def f(self) -> int:
        self.f_runs += 1
        return 0

    def g(self, x: int) -> int:
        self.g_runs += 1
        return 42
```

If you do a walrus like so:

```python
if (x := probe.f()) and (y := probe.g(x)):
    return
```

you may know that `g` won't run since `x` was falsy.

Yet when skimming through code, I read it as:

```python
x = f()
y = g(x)
if x and y:
    return
```

which is wrong, it really should be transformed into: 

```python
x = f()
if not x:
    return

y = g(x)
if not y:
    return
```

You can argue that the shorter variant is better because of its brevity. Yet as someone who reads code more than write it, I take the longer form any day.
